### PR TITLE
[BC-BREAKING] Changed tensor comparison return type from uint8 to bool

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22261 Updated docs and added deprecation warnings to acknowledge a bool tensor
* **#21113 [BC-BREAKING] Changed tensor comparison return type from uint8 to bool**

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

### CODE REVIEW

**Comment by ezyang** [comment]
This is the big kahuna diff, right?!

File: aten/src/ATen/native/Itertools.cpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ Tensor combinations(const Tensor& self, int64_t r, bool with_replacement) {
   std::vector<Tensor> grids = at::meshgrid(std::vector<Tensor>(r, self));
   Tensor mask = _triu_mask(num_elements, r, with_replacement, self.options());
   for(Tensor &t : grids) {
-    t = t.masked_select(mask);
+    t = t.masked_select(mask.to(at::kByte));
```
Comment (ezyang) [inline aten/src/ATen/native/Itertools.cpp]: Going to fix this later?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -126,7 +126,7 @@ Tensor matrix_rank(const Tensor& self, double tol, bool symmetric) {
            "of floating types");
 
   Tensor S = _matrix_rank_helper(self, symmetric);
-  return (S > tol).sum();
+  return (S > tol).to(at::kByte).sum();
```
Comment (ezyang) [inline aten/src/ATen/native/LinearAlgebra.cpp]: `torch.sum` takes a desired dtype, right? Would be better to provide that instead, which saves you from having to do this conversion, and then do another promotion (I am assuming there's a promotion going on here, but I haven't checked...)

Comment (ezyang) [reply aten/src/ATen/native/LinearAlgebra.cpp]: I'm going to call this Note "Sum into dtype" and just reference this in other locations I see it.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -136,7 +136,7 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
 
   Tensor S = _matrix_rank_helper(self, symmetric);
   double tol = _get_epsilon(self.scalar_type()) * std::max(self.size(0), self.size(1));
-  return (S > S.max().mul_(tol)).sum();
+  return (S > S.max().mul_(tol)).to(at::kByte).sum();
```
Comment (ezyang) [inline aten/src/ATen/native/LinearAlgebra.cpp]: Note "Sum into dtype"

File: aten/src/ATen/native/Itertools.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/Itertools.cpp]: I'm going to call this Note "masked_select does not accept bool yet" and reference this at other locations I see it

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -435,7 +435,7 @@ Tensor& logsumexp_out(Tensor& result, const Tensor &self, IntArrayRef dims, bool
   if (self.numel() != 0) {
     auto maxes = at::max_values(self, dims, true);
     auto maxes_squeezed = (keepdim ? maxes : squeeze_multiple(maxes, dims));
-    maxes_squeezed.masked_fill_(maxes_squeezed.abs() == INFINITY, 0);
+    maxes_squeezed.masked_fill_((maxes_squeezed.abs() == INFINITY).to(at::kByte), 0);
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOps.cpp]: Note "masked functions do not accept bool yet"

File: aten/src/ATen/native/Itertools.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/Itertools.cpp]: Wouldn't it make sense to first do all non-BC breaking bool changes first (e.g., make `masked_select` accept bool along with byte), before making the BC-breaking bool changes (making comparison ops return bool)

File: aten/src/ATen/test/atest.cpp [hunk 5]
```diff
@@ -36,7 +36,7 @@ TEST(atest, atest) {
   float b = a.to<float>();
   ASSERT_EQ(b, 4);
 
-  foo = (foo * foo) == (foo.pow(3));
+  foo = ((foo * foo) == (foo.pow(3))).to(kByte);
```
Comment (ezyang) [inline aten/src/ATen/test/atest.cpp]: goofy but this test doesn't make sense anyway, no problems here.

File: test/expect/TestScript.test_listconstruct_erasure.expect [hunk 6]
```diff
@@ -12,7 +12,7 @@ ModelProto {
         Node {type: "Constant", inputs: [], outputs: [1], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
         Node {type: "Less", inputs: [0,1], outputs: [2], attributes: []},
         Node {type: "Cast", inputs: [2], outputs: [3], attributes: [{ name: 'to', type: int, value: 2}]},
-        Node {type: "Cast", inputs: [3], outputs: [4], attributes: [{ name: 'to', type: int, value: 2}]},
+        Node {type: "Cast", inputs: [3], outputs: [4], attributes: [{ name: 'to', type: int, value: 9}]},
```
Comment (ezyang) [inline test/expect/TestScript.test_listconstruct_erasure.expect]: Do we know how this is going to affect saved JIT model backwards compatibility?

File: test/test_distributed.py [hunk 7]
```diff
@@ -645,7 +645,7 @@ def _test_broadcast_helper(
                         tensor = tensor.cuda(rank_to_GPU[rank][0])
                     dist.broadcast(tensor, src, group_id)
                     self.assertEqual(tensor.size(), expected_tensor.size())
-                    self.assertEqual(tensor.ne(expected_tensor).max(), 0)
+                    self.assertEqual(tensor.ne(expected_tensor).max(), torch.tensor(False, dtype=torch.bool))
```
Comment (ezyang) [inline test/test_distributed.py]: I feel like passing `False` here should be OK. But I guess this is testing code so it doesn't matter too much.

File: test/test_distributions.py [hunk 8]
```diff
@@ -2243,8 +2243,8 @@ def test_beta_underflow(self):
             self.assertEqual((beta_samples == 0).sum(), 0)
             self.assertEqual((beta_samples == 1).sum(), 0)
             # assert support is concentrated around 0 and 1
-            frac_zeros = float((beta_samples < 0.1).sum()) / num_samples
-            frac_ones = float((beta_samples > 0.9).sum()) / num_samples
+            frac_zeros = float((beta_samples < 0.1).byte().sum()) / num_samples
+            frac_ones = float((beta_samples > 0.9).byte().sum()) / num_samples
```
Comment (ezyang) [inline test/test_distributions.py]: Note "Sum into dtype"

File: test/test_distributions.py [hunk 9]
```diff
@@ -2254,11 +2254,11 @@ def test_beta_underflow_gpu(self):
         num_samples = 50000
         conc = torch.tensor(1e-2, dtype=torch.float64).cuda()
         beta_samples = Beta(conc, conc).sample([num_samples])
-        self.assertEqual((beta_samples == 0).sum(), 0)
-        self.assertEqual((beta_samples == 1).sum(), 0)
+        self.assertEqual((beta_samples == 0).byte().sum(), 0)
+        self.assertEqual((beta_samples == 1).byte().sum(), 0)
```
Comment (ezyang) [inline test/test_distributions.py]: Note "Sum into dtype"

File: test/test_distributions.py [hunk 10]
```diff
@@ -2254,11 +2254,11 @@ def test_beta_underflow_gpu(self):
...
         # assert support is concentrated around 0 and 1
-        frac_zeros = float((beta_samples < 0.1).sum()) / num_samples
-        frac_ones = float((beta_samples > 0.9).sum()) / num_samples
+        frac_zeros = float((beta_samples < 0.1).byte().sum()) / num_samples
+        frac_ones = float((beta_samples > 0.9).byte().sum()) / num_samples
```
Comment (ezyang) [inline test/test_distributions.py]: Note "Sum into dtype"

File: test/test_indexing.py [hunk 11]
```diff
@@ -86,7 +86,7 @@ def test_multiple_byte_mask(self):
     def test_byte_mask2d(self):
         v = torch.randn(5, 7, 3)
         c = torch.randn(5, 7)
-        num_ones = (c > 0).sum()
+        num_ones = (c > 0).byte().sum()
```
Comment (ezyang) [inline test/test_indexing.py]: Note "Sum into dtype"

File: torch/_tensor_str.py [hunk 12]
```diff
@@ -84,7 +84,7 @@ def __init__(self, tensor):
                 self.max_width = max(self.max_width, len(value_str))
 
         else:
-            nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
+            nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view).byte() & tensor_view.ne(0).byte())
```
Comment (ezyang) [inline torch/_tensor_str.py]: Note "masked functions do not accept bool yet"

File: torch/csrc/jit/symbolic_script.cpp [hunk 13]
```diff
@@ -776,7 +776,7 @@ const std::vector<std::string> functions = {
             other_size = other.size()
             def backward(grad_output):
                 grad_self = (grad_output * condition.type_as(grad_output))._grad_sum_to_size(self_size)
-                grad_other = (grad_output * (1 - condition).type_as(grad_output))._grad_sum_to_size(other_size)
+                grad_other = (grad_output * (1 - condition.byte()).type_as(grad_output))._grad_sum_to_size(other_size)
```
Comment (ezyang) [inline torch/csrc/jit/symbolic_script.cpp]: Shouldn't we just use logical negation here?

File: torch/distributions/constraints.py [hunk 14]
```diff
@@ -269,9 +269,9 @@ class _LowerCholesky(Constraint):
     """
     def check(self, value):
         value_tril = value.tril()
-        lower_triangular = (value_tril == value).view(value.shape[:-2] + (-1,)).min(-1)[0]
+        lower_triangular = (value_tril == value).byte().view(value.shape[:-2] + (-1,)).min(-1)[0]
```
Comment (ezyang) [inline torch/distributions/constraints.py]: A better fix is to change `min(-1)` into `any(-1)`, no?

File: torch/distributions/constraints.py [hunk 15]
```diff
@@ -269,9 +269,9 @@ class _LowerCholesky(Constraint):
...
 
-        positive_diagonal = (value.diagonal(dim1=-2, dim2=-1) > 0).min(-1)[0]
+        positive_diagonal = (value.diagonal(dim1=-2, dim2=-1) > 0).byte().min(-1)[0]
```
Comment (ezyang) [inline torch/distributions/constraints.py]: Ditto here

File: torch/distributions/fishersnedecor.py [hunk 16]
```diff
@@ -50,13 +50,13 @@ def expand(self, batch_shape, _instance=None):
     @property
     def mean(self):
         df2 = self.df2.clone()
-        df2[df2 <= 2] = nan
+        df2[(df2 <= 2).byte()] = nan
```
Comment (ezyang) [inline torch/distributions/fishersnedecor.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/fishersnedecor.py [hunk 17]
```diff
@@ -50,13 +50,13 @@ def expand(self, batch_shape, _instance=None):
...
         return df2 / (df2 - 2)
 
     @property
     def variance(self):
         df2 = self.df2.clone()
-        df2[df2 <= 4] = nan
+        df2[(df2 <= 4).byte()] = nan
```
Comment (ezyang) [inline torch/distributions/fishersnedecor.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/fishersnedecor.py (hunk 16 above)
Comment (ezyang) [reply torch/distributions/fishersnedecor.py]: Actually, breakage of this code is very serious, and I think a very strong argument that you have to fix the masked functions first before landing this diff.

File: torch/distributions/geometric.py [hunk 18]
```diff
@@ -89,7 +89,7 @@ def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
         value, probs = broadcast_all(value, self.probs.clone())
-        probs[(probs == 1) & (value == 0)] = 0
+        probs[(probs == 1).byte() & (value == 0).byte()] = 0
```
Comment (ezyang) [inline torch/distributions/geometric.py]: Note "masked functions do not accept bool yet"... maybe???

File: torch/distributions/half_cauchy.py [hunk 19]
```diff
@@ -50,7 +50,7 @@ def variance(self):
 
     def log_prob(self, value):
         log_prob = self.base_dist.log_prob(value) + math.log(2)
-        log_prob[value.expand(log_prob.shape) < 0] = -inf
+        log_prob[(value.expand(log_prob.shape) < 0).byte()] = -inf
```
Comment (ezyang) [inline torch/distributions/half_cauchy.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/half_normal.py [hunk 20]
```diff
@@ -50,7 +50,7 @@ def variance(self):
 
     def log_prob(self, value):
         log_prob = self.base_dist.log_prob(value) + math.log(2)
-        log_prob[value.expand(log_prob.shape) < 0] = -inf
+        log_prob[(value.expand(log_prob.shape) < 0).byte()] = -inf
```
Comment (ezyang) [inline torch/distributions/half_normal.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/kl.py [hunk 21]
```diff
@@ -177,11 +177,11 @@ def kl_divergence(p, q):
 @register_kl(Bernoulli, Bernoulli)
 def _kl_bernoulli_bernoulli(p, q):
     t1 = p.probs * (p.probs / q.probs).log()
-    t1[q.probs == 0] = inf
-    t1[p.probs == 0] = 0
+    t1[(q.probs == 0).byte()] = inf
+    t1[(p.probs == 0).byte()] = 0
```
Comment (ezyang) [inline torch/distributions/kl.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/kl.py [hunk 22]
```diff
@@ -177,11 +177,11 @@ def kl_divergence(p, q):
...
     t2 = (1 - p.probs) * ((1 - p.probs) / (1 - q.probs)).log()
-    t2[q.probs == 1] = inf
-    t2[p.probs == 1] = 0
+    t2[(q.probs == 1).byte()] = inf
+    t2[(p.probs == 1).byte()] = 0
```
Comment (ezyang) [inline torch/distributions/kl.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/kl.py [hunk 23]
```diff
@@ -204,7 +204,7 @@ def _kl_binomial_binomial(p, q):
     if (p.total_count < q.total_count).any():
         raise NotImplementedError('KL between Binomials where q.total_count > p.total_count is not implemented')
     kl = p.total_count * (p.probs * (p.logits - q.logits) + (-p.probs).log1p() - (-q.probs).log1p())
-    inf_idxs = p.total_count > q.total_count
+    inf_idxs = (p.total_count > q.total_count).byte()
```
Comment (ezyang) [inline torch/distributions/kl.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/kl.py [hunk 24]
```diff
@@ -411,7 +411,7 @@ def _kl_pareto_pareto(p, q):
     t1 = q.alpha * scale_ratio.log()
     t2 = -alpha_ratio.log()
     result = t1 + t2 + alpha_ratio - 1
-    result[p.support.lower_bound < q.support.lower_bound] = inf
+    result[(p.support.lower_bound < q.support.lower_bound).byte()] = inf
```
Comment (ezyang) [inline torch/distributions/kl.py]: Note "masked functions do not accept bool yet"

File: torch/distributions/kl.py [hunk 25]
```diff
@@ -481,7 +481,7 @@ def _kl_beta_normal(p, q):
 @register_kl(Beta, Uniform)
 def _kl_beta_uniform(p, q):
     result = -p.entropy() + (q.high - q.low).log()
-    result[(q.low > p.support.lower_bound) | (q.high < p.support.upper_bound)] = inf
+    result[(q.low > p.support.lower_bound).byte() | (q.high < p.support.upper_bound).byte()] = inf
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/distributions/kl.py]: I'm going to stop labeling these, there are a lot of these sites.

**Comment by ezyang** [comment]
If any of the BC-breaking changes are intended to be durable, I expect to see detailed information in the commit message about how to fix old code with the old behavior.

File: torch/distributions/kl.py (hunk 24 above)
Comment (izdeby) [reply torch/distributions/kl.py]: Masked functions and indexing are supported for bool masks.
One of the previous PRs in this stack fixed this issue - indexing on scalars.

**Comment by ezyang** [comment]
OK, in that case, are these changes necessary?

Excerpts from iurii zdebskyi's message of 2019-06-03 09:11:10 -0700:
> izdeby commented on this pull request.
> 
> > @@ -411,7 +411,7 @@ def _kl_pareto_pareto(p, q):
>      t1 = q.alpha * scale_ratio.log()
>      t2 = -alpha_ratio.log()
>      result = t1 + t2 + alpha_ratio - 1
> -    result[p.support.lower_bound < q.support.lower_bound] = inf
> +    result[(p.support.lower_bound < q.support.lower_bound).byte()] = inf
> 
> Masked functions and indexing are supported for bool masks.
> One of the previous PRs in this stack fixed this issue - indexing on scalars.
>

**Comment by ezyang** [comment]
Much better! But I was under the impression that sums on uint8 promoted into int64 previously; if that's the case, `ScalarType::Int` is the wrong dtype to use.

**Review by ezyang — CHANGES_REQUESTED** [review]
Either accumulate into int64, or add a prominent comment saying that we've changed the return precision of some operations from int64 to int32

File: test/test_torch.py [hunk 26]
```diff
@@ -9186,7 +9186,7 @@ def _test_bernoulli(self, t_dtype, p_dtype, device):
             self.assertEqual(x.bernoulli().tolist(), trivial_p)
 
         def isBinary(t):
-            return torch.ne(t, 0).mul_(torch.ne(t, 1)).sum().item() == 0
+            return (torch.ne(t, 0) == torch.ne(t, 1)).all().item() == False
```
Comment (ezyang) [inline test/test_torch.py]:
```
./test/test_torch.py:9189:68: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
```

**Review by ezyang — APPROVED** [review]
please fix lint before landing

**Comment by ezyang** [comment]
And, for the record, we're going to subsequently change `sum(booltensor)` to accumulate into `int64` by default, thus resolving the BC-breakage that is fixed a few lines here.

File: aten/src/ATen/Declarations.cwrap [hunk 27]
```diff
@@ -364,6 +364,7 @@
 [[
   name: _th_equal
   cpu_bool: True
+  cuda_bool: True
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: this is orthogonal right?  I.e. can be in a separate PR, there's nothing that has to change as a result of comparison op return types changing.

File: aten/src/ATen/Declarations.cwrap [hunk 28]
```diff
@@ -486,6 +487,8 @@
 [[
   name: _th_ixor_
   cname: __ixor__
+  cpu_bool: True
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: same here?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/Declarations.cwrap (hunk 27 above)
Comment (izdeby) [reply aten/src/ATen/Declarations.cwrap]: https://github.com/pytorch/pytorch/pull/21538

**Review by VitalyFedyunin — APPROVED** [review]

File: tools/autograd/templates/python_variable_methods.cpp [hunk 29]
```diff
@@ -291,8 +291,8 @@ static Tensor dispatch_invert(const Tensor & self) {
 static PyObject * THPVariable_invert(PyObject* self, PyObject* args) {
   HANDLE_TH_ERRORS
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-  if (self_.scalar_type() != at::kByte) {
-    throw TypeError("~ (operator.invert) is only implemented on byte tensors");
+  if (self_.scalar_type() != at::kByte && self_.scalar_type() != at::kBool) {
```
[the remark this answers is not in the document]
Comment (gchanan) [reply tools/autograd/templates/python_variable_methods.cpp]: ok, can you file an issue?  This looks questionable because it would seem the JIT would have to implement the same logic for no good reason.

**Review by gchanan — CHANGES_REQUESTED** [review]
I would expect more docs to change?  For example, we have docs for `torch.gt` that look wrong after this change.

File: torch/csrc/jit/symbolic_script.cpp (hunk 13 above)
Comment (gchanan) [reply torch/csrc/jit/symbolic_script.cpp]: are you tracking this somewhere in an issue?

**Review by apaszke — COMMENTED** [review]
This also requires an update to the shape inference code in the JIT

**Comment by izdeby** [comment]
> This also requires an update to the shape inference code in the JIT
hey, @apaszke, im not very familiar with the JIT codebase yet, can you, please, point me to a place which needs to be updated? Would you like it to be part of this PR or a separate as a prep work?

**Comment by izdeby** [comment]
> I would expect more docs to change? For example, we have docs for `torch.gt` that look wrong after this change.

@gchanan, here is the docs [pr](https://github.com/pytorch/pytorch/pull/22261)

**Comment by apaszke** [comment]
@izdeby it needs to happen inside this PR, or the JIT will be broken until it is fixed. Fortunately I think the change is as simple as updating [this line](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/shape_analysis.cpp#L960) to the bool scalar type (if that's a thing? I haven't followed the bool tensor changes).

File: tools/autograd/templates/python_variable_methods.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (izdeby) [reply tools/autograd/templates/python_variable_methods.cpp]: https://github.com/pytorch/pytorch/issues/22262

**Comment by izdeby** [comment]
@apaszke, thanks for pointing this out. im surprised that nothing failed even without me changing the JIT part. is there any way i can test it besides the test_jit.py (which passes before and after the change)?

**Comment by apaszke** [comment]
`test_jit.py` should cover that. We should open an issue about this

File: aten/src/ATen/native/Itertools.cpp [hunk 30]
```diff
@@ -12,7 +12,7 @@ Tensor _triu_mask(int64_t n, int64_t dims, bool diagonal, TensorOptions opt) {
   // or i <= j <= k <= ... (depending on diagonal)
   Tensor range = at::arange(n, opt.dtype(kLong));
   std::vector<Tensor> index_grids = at::meshgrid(std::vector<Tensor>(dims, range));
-  Tensor mask = at::ones(index_grids[0].sizes(), opt.dtype(kByte));
+  Tensor mask = at::ones(index_grids[0].sizes(), opt.dtype(kBool));
```
Comment (gchanan) [inline aten/src/ATen/native/Itertools.cpp]: I wonder (you don't have to make this change): does `at::full(..., True, opt.dtype(kBool));` work?  That way you avoid specifying ones for bool.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 31]
```diff
@@ -136,7 +136,7 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
 
   Tensor S = _matrix_rank_helper(self, symmetric);
   double tol = _get_epsilon(self.scalar_type()) * std::max(self.size(0), self.size(1));
-  return (S > S.max().mul_(tol)).sum();
+  return (S > S.max().mul_(tol)).sum(ScalarType::Long);
```
Comment (gchanan) [inline aten/src/ATen/native/LinearAlgebra.cpp]: do you actually need this?  Isn't this the default now?

File: c10/core/ScalarType.h [hunk 32]
```diff
@@ -109,7 +109,8 @@ namespace c10 {
   _(double, Double, d)                        \
   _(c10::qint8, QInt8, i)                     \
   _(c10::quint8, QUInt8, i)                   \
-  _(c10::qint32, QInt32, i)
+  _(c10::qint32, QInt32, i)                   \
+  _(bool, Bool, i)
```
Comment (gchanan) [inline c10/core/ScalarType.h]: this doesn't seem consistent (some of the FORALLs have bool and some dont).  What makes this special?

File: torch/csrc/jit/symbolic_script.cpp (hunk 13 above)
Comment (gchanan) [reply torch/csrc/jit/symbolic_script.cpp]: did you answer this?

**Review by gchanan — APPROVED** [review]
I'm a bit concerned with the FORALL macros, but okay.

File: c10/core/ScalarType.h (hunk 32 above)
Comment (izdeby) [reply c10/core/ScalarType.h]: As of right now, there are only 2 FORALL marcos that dont have Bool in it.
1) AT_FORALL_SCALAR_TYPES_EXCEPT_QINT
Its being used only in core/Scalar.h for conversion and there is a bool overload because of a compiler error. Please see a comment [there](https://github.com/pytorch/pytorch/blob/master/c10/core/Scalar.h#L38). 

2) AT_FORALL_SCALAR_TYPES_EXCEPT_HALF_AND_QINT()
I didn't touch it as there was never a real need to. its being used for in
- THBlasUtils.h 
- TensorFactories.cpp

In order to keep this PR clean, i created an [issue](https://github.com/pytorch/pytorch/issues/22744) to figure out what to do with the AT_FORALL_SCALAR_TYPES_EXCEPT_HALF_AND_QINT macro.

**Comment by colesbury** [comment]
You may need to include part of https://github.com/pytorch/pytorch/pull/22326 in this PR. Basically, something like `~(tensor > 2)` should continue to work to minimize the impact of breaking backwards compatibility.

**Comment by izdeby** [comment]
> You may need to include part of #22326 in this PR. Basically, something like `~(tensor > 2)` should continue to work to minimize the impact of breaking backwards compatibility.

Im using it this PR :) in torch/csrc/jit/symbolic_script.cpp

**Comment by colesbury** [comment]
@izdeby, cool. Does it also continue to work in Python?

(In case it's not clear I had to revert https://github.com/pytorch/pytorch/pull/22326 because it changed the uint8 behavior while comparisons were still returning uint8)

**Comment by ailzhang** [comment]
cc: @dlibenzi @asuhan This requires a followup change on XLA side once landed.

**Comment by dlibenzi** [comment]
> cc: @dlibenzi @asuhan This requires a followup change on XLA side once landed.

Thanks @ailzhang !
I think it makes sense to give a structure to those warnings.
Like, create pytorch_xla issues referencing the pytorch PR which will be causing the issue.

**Comment by dlibenzi** [comment]
> > cc: @dlibenzi @asuhan This requires a followup change on XLA side once landed.
> 
> Thanks @ailzhang !
> I think it makes sense to give a structure to those warnings.
> Like, create pytorch_xla issues referencing the pytorch PR which will be causing the issue.

Created https://github.com/pytorch/xla/issues/815

**Comment by ailzhang** [comment]
@dlibenzi That's a good idea! Going forward I'm going to create issues for those that I think need XLA followup changes. 
I will only cc you/Alex for issues that I'm not certain whether it's the simple change or not. What do you think?

**Comment by gchanan** [comment]
Is symbolic_variable used anywhere?

https://github.com/pytorch/pytorch/blob/9eb039334f2b84e61143363ce7f9a0aa7d6dfd2c/torch/csrc/jit/symbolic_variable.h#L81-L120 all looks wrong.

**Comment by ailzhang** [comment]
@gchanan I think those are not in use, but good to update them for now. I had a WIP PR before aiming to remove symbolic_variable in JIT codebase, but didn't finish it due to other high pri items. I can get back to it depends on the priority.

**Comment by gchanan** [comment]
one other note: there isn't any BC support for passing uint8 tensors to `out=` for ne, eq, etc.

File: aten/src/TH/generic/THTensorMoreMath.cpp [hunk 33]
```diff
@@ -9,30 +9,30 @@
...
-    THByteTensor_resizeNd(r_, ta->dim(), THTensor_getSizePtr(ta), NULL); \
-    TH_TENSOR_APPLY3(unsigned char, r_, scalar_t, ta, scalar_t, tb, \
-                     *r__data = (*ta_data OP *tb_data) ? 1 : 0;); \
-  } \
-  void THTensor_(NAME##TensorT)(THTensor *r_, THTensor *ta, THTensor *tb) \
-  { \
-    THTensor_(resizeNd)(r_, ta->dim(), THTensor_getSizePtr(ta), NULL); \
-    TH_TENSOR_APPLY3(scalar_t, r_, scalar_t, ta, scalar_t, tb, \
-                     *r__data = (*ta_data OP *tb_data) ? 1 : 0;); \
+#define TENSOR_IMPLEMENT_LOGICAL(NAME,OP)                                      \
```
Comment (colesbury) [inline aten/src/TH/generic/THTensorMoreMath.cpp]: I know this predates your change, but none of the instantiations of this macro are logical operators (they're relational operators). The logical operators in C are `!, &&, ||`

File: test/test_torch.py [hunk 34]
```diff
@@ -11366,9 +11402,20 @@ def test_bitwise_ops(self):
         x_clone ^= y
         self.assertEqual(x_clone, xor_result)
 
-    def test_invert(self):
-        x = torch.ByteTensor([0, 1, 1])
-        self.assertEqual((~x).tolist(), [1, 0, 0])
+    def test_op_invert(self):
+        res = 0xffff - torch.arange(127, dtype=torch.int8)
```
Comment (gchanan) [inline test/test_torch.py]: what are you trying to test here?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cuda/LegacyDefinitions.cpp [hunk 35]
```diff
@@ -52,4 +52,123 @@ Tensor gather_cuda(const Tensor & self, int64_t dim, const Tensor & index, bool
   return legacy::cuda::_th_gather(self, dim, index);
 }
 
+Tensor & lt_out_cuda(Tensor & result, const Tensor & self, const Tensor & other) {
+  if (result.dtype() == at::ScalarType::Byte) {
+    AT_WARN("torch.lt received a result with dtype torch.uint8, this behavior is now deprecated," \
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/native/cuda/LegacyDefinitions.cpp]: `out parameter`

File: test/test_torch.py (hunk 34 above)
Comment (gchanan) [reply test/test_torch.py]: I see this was just taken from https://github.com/pytorch/pytorch/pull/22320/ so you can leave it.  It does look strange though --e.g. `0xffff - torch.arange(127, dtype=torch.int8)` might not even give you an int8 once we have type promotion.

**Comment by izdeby** [comment]
@gchanan, regarding 

> did you answer this question?

 i cant replace at::ones with at::full as first will return `[ Variable[CPUBoolType]{SIZE} ]` and second will give `[ CPUDoubleType{SIZE} ]`

File: aten/src/ATen/native/Itertools.cpp (hunk 30 above)
Comment (izdeby) [reply aten/src/ATen/native/Itertools.cpp]: i cant replace at::ones with at::full as first will return [ Variable[CPUBoolType]{SIZE} ] and second will give [ CPUDoubleType{SIZE} ]

Comment (gchanan) [reply aten/src/ATen/native/Itertools.cpp]: I think this is because you still need to pass dtype bool.

**Review by gchanan — APPROVED** [review]
make the one `full` change and this looks good.

File: torch/onnx/symbolic_opset8.py [hunk 36]
```diff
@@ -98,12 +98,12 @@ def _comparison_operator(g, input, other, op_name):
 
 # NOTE: For symbolics {gt, lt, bmm, matmul, prelu, mm, addmm, view, flatten},
 #       integer input type not supported in opset8. Cast to float if possible.
-@wrap_logical_op_with_cast_to('Byte')
```
Comment (izdeby) [inline torch/onnx/symbolic_opset8.py]: @houseroad

File: torch/onnx/symbolic_opset9.py [hunk 37]
```diff
@@ -769,18 +769,18 @@ def wrap_with_not(g, input, other):
     return wrap_with_not
 
 
-@wrap_logical_op_with_cast_to('Byte')
```
Comment (izdeby) [inline torch/onnx/symbolic_opset9.py]: @houseroad

Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: Should be ok to remove these `@wrap_logical_op_with_cast_to('Bool')` decorators. These ONNX ops return output with dtype Bool. They were added to be compatible with PyTorch (uint8).

File: test/onnx/test_pytorch_onnx_caffe2.py [hunk 38]
```diff
@@ -1398,7 +1398,7 @@ def forward(self, x):
     def test_isnan(self):
         class IsNaNModel(torch.nn.Module):
             def forward(self, input):
-                return torch.isnan(input)
+                return torch.isnan(input).byte()
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_caffe2.py]: instead of add `.byte()`, remove the line https://github.com/pytorch/pytorch/blob/b22adeb0074f57c501422a6f44c5e0ad6f1c4adb/torch/onnx/symbolic_opset9.py#L1602

File: torch/onnx/symbolic_opset8.py (hunk 36 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset8.py]: same as below

File: test/onnx/test_pytorch_onnx_caffe2.py (hunk 38 above)
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_caffe2.py]: do we still need this conversion if we remove the cast from symbolic?

File: torch/onnx/symbolic_opset9.py (hunk 37 above)
Comment (houseroad) [reply torch/onnx/symbolic_opset9.py]: I think @BowenBao is right. Also I am wondering whether we should keep wrap_logical_op_with_cast_to function anymore, since it should be used anywhere if I remember correctly.

**Review by houseroad — COMMENTED** [review]
Aligning the logic of symbolic to PyTorch is the right action. Please make sure the test is green.

**Review by BowenBao — COMMENTED** [review]
Added few comments on the ONNX exporter side. Also how would this change affect mask related operators? Like `masked_fill` that takes input `mask` of `ByteTensor`

**Comment by izdeby** [comment]
@ailzhang, this PR breaks XLA.

**Comment by ailzhang** [comment]
@izdeby  Thanks! Alex had https://github.com/pytorch/xla/pull/833 ready to mitigate the breakage once this lands. :D

**Comment by dlibenzi** [comment]
> @izdeby Thanks! Alex had [pytorch/xla#833](https://github.com/pytorch/xla/pull/833) ready to mitigate the breakage once this lands. :D

If XLA is the only broken test, you can go in, and we will follow with that PR.

**Comment by ezyang** [comment]
Lint seems to be broken by this PR
